File: Coarse5/ch01/java-example/src/main/java/org/example/ex02/Exam02.java
package org.example.ex02;

public class Exam02 {

    private int a;

    public Exam02(){
        var b = 20; // << 타입 추론
//        var c = null;
        int c = 30;
        int d;
        Integer e = new Integer(100);
        Integer f = 20;
        Integer g = null;

        callFunction(a);
        callFunction(b);
        callFunction(c);
//        callFunction(d); // 초기화가 안되었기 때문에 에러 발생.
        callFunction(e); // null 값이기 때문에 에러 발생.
        callFunction(f);
        callFunction(g);

        // 자바에서는 값이 null인지 아닌지에 따라 개발자가 신경써서 처리해주어야 한다.
    }

    public void callFunction(Integer i){

        var temp = (i==null) ? Integer.valueOf(100) : Integer.valueOf(i);
//        var temp = (i==null) ? "null 값 입니다" : i;
        System.out.println(temp);
    }

    public static void main(String[] agrs){
        new Exam02();
    }
}

File: Coarse5/ch01/kotlin-example/src/main/kotlin/ex02/Exam02.kt
package org.example.ex02

import java.util.*

fun main(){
    val a:Int = 0
    val b = 10
    val c: Int = 20
    val d: Int? = null

    callFunction(a)
    callFunction(b)
    callFunction(c)
    callFunction(d) // null 을 출력해준다.
    callFunction()
//    Optional.ofNullable(d).ifPresent{it -> println(it) }
}

// ? : 엘비스 연산자(null일수도 있다)
// 변수? << 변수가 null인가?
// 변수?. << 변수가 null이 아닐 때
// 변수?: << 변수가 null일 때
fun callFunction(i: Int? = 100){ // 넘어오는 값이 없으면 디폴트 값 100으로 설정.

    // 1
//    i?.let {
//        println(it)
//    }?: run {
//        println("null 값 입니다.")
//    }

    // 2
//    val temp = if(i==null) "null 값 입니다" else i
//    println(temp)

    // 3
    val temp = i ?: "null 값 입니다"
    println(temp)

    // 4
    val temp2 = i?.let { it * 20 } ?: "null 값 입니다"
    println(temp2)
}